src/ripple/app/reporting/ETLHelpers.h

```diff
@@ -0,0 +1,182 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_REPORTING_ETLHELPERS_H_INCLUDED
+#define RIPPLE_APP_REPORTING_ETLHELPERS_H_INCLUDED
+#include <ripple/app/main/Application.h>
+#include <ripple/ledger/ReadView.h>
+#include <condition_variable>
+#include <mutex>
+#include <optional>
+#include <queue>
+#include <sstream>
+
+namespace ripple {
+
+/// This datastructure is used to keep track of the sequence of the most recent
+/// ledger validated by the network. There are two methods that will wait until
+/// certain conditions are met. This datastructure is able to be "stopped". When
+/// the datastructure is stopped, any threads currently waiting are unblocked.
+/// Any later calls to methods of this datastructure will not wait. Once the
+/// datastructure is stopped, the datastructure remains stopped for the rest of
+/// its lifetime.
+class NetworkValidatedLedgers
+{
+    // max sequence validated by network
+    std::optional<uint32_t> max_;
+
+    mutable std::mutex m_;
+
+    std::condition_variable cv_;
+
+    bool stopping_ = false;
+
+public:
+    /// Notify the datastructure that idx has been validated by the network
+    /// @param idx sequence validated by network
+    void
+    push(uint32_t idx)
+    {
+        std::lock_guard lck(m_);
+        if (!max_ || idx > *max_)
+            max_ = idx;
+        cv_.notify_all();
+    }
+
+    /// Get most recently validated sequence. If no ledgers are known to have
+    /// been validated, this function waits until the next ledger is validated
+    /// @return sequence of most recently validated ledger. empty optional if
+    /// the datastructure has been stopped
+    std::optional<uint32_t>
+    getMostRecent()
+    {
+        std::unique_lock lck(m_);
+        cv_.wait(lck, [this]() { return max_ || stopping_; });
+        return max_;
+    }
+
+    /// Waits for the sequence to be validated by the network
+    /// @param sequence to wait for
+    /// @return true if sequence was validated, false otherwise
+    /// a return value of false means the datastructure has been stopped
+    bool
+    waitUntilValidatedByNetwork(uint32_t sequence)
+    {
+        std::unique_lock lck(m_);
+        cv_.wait(lck, [sequence, this]() {
+            return (max_ && sequence <= *max_) || stopping_;
+        });
+        return !stopping_;
+    }
+
+    /// Puts the datastructure in the stopped state
+    /// Future calls to this datastructure will not block
+    /// This operation cannot be reversed
+    void
+    stop()
+    {
+        std::lock_guard lck(m_);
+        stopping_ = true;
+        cv_.notify_all();
+    }
+};
+
+/// Generic thread-safe queue with an optional maximum size
+template <class T>
+class ThreadSafeQueue
```

Comment:
`boost::lockfree:queue` behaves differently than `ThreadSafeQueue`. `ThreadSafeQueue` provides a blocking API, where we can wait for the next element to be inserted, or wait for free space. Unless I'm reading the documentation wrong, `boost::lockfree::queue` does not provide either of these waiting functionalities. Waiting is precisely what the application needs to do, so if we used `boost::lockfree::queue`, we would still need to implement the waiting ourselves via condition variables (or busy waiting). I think this would be more complex than just using `ThreadSafeQueue`. In fact, if we used `boost::lockfree::queue`, we would most likely end up just replacing the `std::queue` member with `boost::lockfree::queue`, since we would still need the waiting semantics that `ThreadSafeQueue` class provides.

Reply:
That's fine. Thanks for the exposition. You may want to add a comment on _why_ we need a blocking queue and a lockfree one won't do.